Add 362 release notes (#9216)
trinodb/trino: docs/src/main/sphinx/release/release-362.md

  supported by Trino. Writing to such tables is still not allowed. ({issue}`8986`)
* Extend set of statistics reported by JMX to include metrics for calls made to the Glue statistics API. ({issue}`9100`) 
* Limit maximum file size generated by write operations to 1 GB by default. The limit is not exact and is applied on a best-effort basis. 
  The limit can be set with the `hive.target-max-file-size` configuration property or the `target_max_file_size` session property. ({issue}`7991`) 
* Allow specifying the list of schemas for which Trino will enforce that queries use a filter on partition keys for source tables.
  The list can be specified using the `hive.query-partition-filter-required-schemas`, or the `query_partition_filter_required_schemas` session property.
  The list is taken into consideration only if the `hive.query-partition-filter-required` configuration property or the `query_partition_filter_required` 
  session property is set to `true`. ({issue}`9106`)
* Fix failure when writing Parquet files with Snappy compression on ARM64. ({issue}`9148`)
* Fix performance regression when reading Parquet files that contain dictionaries. ({issue}`9161`)

> * Fix incorrect table locking in Hive metastore when querying transactional and not partitioned table. ({issue}`9070`) 

Reviewer: WDYT `unpartitioned non-transactional table`.

> * Fix `ArrayIndexOutOfBoundsException` when using the experimental Parquet writer. ({issue}`5518`)

Reviewer: Is it experimental or "native"? I'm confused which is which.

Author: It's both, although "native" is not a good characterization -- "optimized" is what we've called it in other contexts. In fact the description for the config option says "Experimental: Enable optimized Parquet writer".

* Fix reading Avro tables written with older Avro versions that had certain semi-invalid schemas. ({issue}`9114`)
* Fix possible `INSERT`/`UPDATE`/`ANALYZE` query failure when Glue metastore is in use and table statistics collection is enabled. ({issue}`9297`)

## Iceberg connector

* Add support for Iceberg `uuid` type. ({issue}`6663`)
* Fix the mapping of nested fields between table metadata and Parquet file metadata. This
  enables evolution of `row` typed columns for Iceberg tables stored in Parquet. ({issue}`9124`)
* Fix failure for queries filtering on columns with array, map, or row type. ({issue}`8822`)

> * Fix `ArrayIndexOutOfBoundsException` when writing Parquet data. ({issue}`5518`)

Reviewer: This seems more broad than it really is (applies only to the experimental writer). Let's use same text as the Hive connector.

``Fix `ArrayIndexOutOfBoundsException` when using the experimental Parquet writer. ({issue}`5518`)``

> * Fix query failures caused by errors reading certain Avro metadata files. ({issue}`9114`)

Reviewer: WDYT? Or is it too much detail?

```suggestion
* Fix query failures caused by errors reading certain Avro metadata files written with older Avro versions that had certain semi-invalid schemas. ({issue}`9114`)
```

Author: I'm not sure it add much value. "older versions" and "semi-invalid" schemas are too vague to be useful.

> # Release 362 (XX Sep 2021)

Reviewer: reminder for the date

## General

> * Add support for {func}`LISTAGG`. ({issue}`4835`)

Reviewer 2: Lowercase `listagg` to enable a reference. 
<img width="374" alt="Screen Shot 2021-09-13 at 0 54 09" src="https://user-images.githubusercontent.com/6237050/132994348-c9672f6d-e7e9-40f1-a16f-879817059095.png">
<img width="353" alt="Screen Shot 2021-09-13 at 0 54 43" src="https://user-images.githubusercontent.com/6237050/132994350-4c7d3556-3c78-45e5-a086-336ee9de51ad.png">

* Improve join performance. ({issue}`8974`)
* Improve performance of queries that contain `DISTINCT` clause. ({issue}`8967`)
* Improve query performance by reducing CPU overhead of repartitioning data across worker nodes. ({issue}`9102`)
* Prevent failure when querying `system.materialized_views` and materialized view is dropped concurrently. ({issue}`9050`) 

> * Fix `ALTER SCHEMA ... SET AUTHORIZATION` to resolve roles using the catalog of the schema instead of the session catalog. ({issue}`9066`)

Reviewer 3: Is this a breaking fix? Are the docs updated?

Author: cc @dain

* Fix query failure when query contains a cast from `varchar` to a shorter `char`. ({issue}`9036`)

> * Fix planning of recursive queries when the recursion the base plan or the recursion step plan produce duplicate outputs.

Reviewer: Should there be a comma here?

```suggestion
* Fix planning of recursive queries when the recursion, the base plan or the recursion step plan produce duplicate outputs.
```

Reviewer 3: needs one more comma.  

the base plan, or the

* Split metrics that are reported in operator stats into `connectorMetrics` and `metrics`. ({issue}`9156`)

## Security

* Add `IN catalog` clause to `CREATE ROLE`, `DROP ROLE`, `GRANT ROLE`, `REVOKE ROLE`, and `SET ROLE` to specify 

>   the target catalog of the statement instead of using the current session catalog. This change is necessary to 

Reviewer 4: Call it out that this is a breaking change.
cc @mosabua

Reviewer 3: Agreed .. also we have to make sure this is actually documented in the docs ... is that the case? If not can whoever implemented this feature send a PR with the relevant docs changes. From the explanation here I am not clear what is going on.

Author: cc @dain

  support system roles in the future. The old behavior of these statements can be restored by setting the 
  `deprecated.legacy-catalog-roles` config property. ({issue}`9087`)

## BigQuery connector

* Support case insensitive name matching for BigQuery views. ({issue}`9164`)

> * Change type mapping of BigQuery `datetime` from `timestamp(3)` to `timestamp(6)` in Trino. ({issue}`9052`)

Reviewer 3: Are the docs for all these type mappings updated?

Reviewer 2: Yes, they're already updated.
https://github.com/trinodb/trino/blob/master/docs/src/main/sphinx/connector/bigquery.rst#data-types

* Change type mapping of BigQuery `time` from `time with time zone` to `time(6)` in Trino. ({issue}`9052`)
* Change type mapping of BigQuery `timestamp` from `timestamp(3) with time zone` to `timestamp(6) with time zone` in Trino. ({issue}`9052`)
* Fix query failure for predicate pushdown to BigQuery `datetime` type. ({issue}`9005`)
* Fix incorrect results when using parameterized `numeric` type with non-default precision and scale. ({issue}`9060`)

> * Fix query failure when accessing tables having unsupported type columns. ({issue}`9086`)

Reviewer: `having columns of unsupported data type`? WDYT?

Reviewer 3: when accessing tables with unsupported data type columns

> * Fix query failure for predicate pushdown to BigQuery `float64` type with `+infinity` or `-infinity` value. ({issue}`9122`)

Reviewer: WDYT?

```suggestion
* Fix query failure when pushing down predicates to BigQuery on `float64` columns with `+infinity` or `-infinity` value. ({issue}`9122`)
```

## Cassandra connector

* Change minimum number of speculative executions from 2 to 1. ({issue}`9096`)

## Hive connector

> * Limit maximum file size to 1 GB by default.  The limit is not exact and is performed best effort. The limit can be set with 

Reviewer: `Maximum file size generated by write operations`? This reads like we will not query files > 1GB.

`The limit is not exact and is applied on a best-effort basis`? WDYT?

>   the `target_max_file_size` session property, or the `hive.target-max-file-size` configuration property. ({issue}`7991`)

Reviewer 3: This should be in the docs .. not here. Also is this just for write or also read?

> * Fix file-based metastore to always add the admin role by default. ({issue}`9066`)

Reviewer 2: If I remember correctly, we haven't documented file-based metastore intentionally. It would be better to remove if unnecessary.

* Fix failure when writing Parquet files with Snappy compression on ARM64. ({issue}`9148`)

> * Fix performance regression where reading Parquet files that contain dictionaries. ({issue}`9161`)

Reviewer: `s/where/when/`?

Were all reads impacted or just the ones that applied predicates?

Author: I didn't look closely, but based on user reports, it appears to affect any queries involving files with dictionary-encoded columns.

## Iceberg connector

* Fix the mapping of nested fields between table metadata and Parquet file metadata. This
  enables evolution of `row` typed columns for Iceberg tables stored in Parquet. ({issue}`9124`)

## Pinot connector

* Support pushdown of filters on `VARBINARY` columns to Pinot. ({issue}`9180`)

> * Fix handling of `IN` and `NOT IN` clauses for `VARCHAR` columns in presence of aggregations. ({issue}`9133`)

Reviewer 2: I feel "Fix handling" is little unclear. It was incorrect result issue?

Reviewer: Yes. I think it was a correctness issue (but pre-existing).

cc: @elonazoulay can you clarify once?

## TPC-H connector

> * Add support for switch mapping for floating point values between SQL `DOUBLE` and `DECIMAL` types. The mapping 
>   can be set  the `tpch.double-type-mapping` configuration property. ({issue}`7990`)

Reviewer: ```suggestion
* Add support to switch mapping of floating point values between SQL `DOUBLE` and `DECIMAL` types. The mapping 
  can be set using the `tpch.double-type-mapping` configuration property. ({issue}`7990`)
```

Author: I think this should be "Add support for switching the mapping of ..."

# Release 362 (XX Sep 2021)

## General

* Add support for {func}`listagg`. ({issue}`4835`)
* Improve join performance. ({issue}`8974`)
* Improve performance of queries that contain `DISTINCT` clause. ({issue}`8967`, {issue}`9194`)
* Improve query performance by reducing CPU overhead of repartitioning data across worker nodes. ({issue}`9102`)

> * Limit number of metadata calls done for particular queries reading from `system.jdbc.columns` where filter on schema name is present. ({issue}`9214`)

Reviewer 5: ```suggestion
* Limit the number of backend SPI metadata calls when `java.sql.DatabaseMetaData` API is used to list table columns with a filter on schema name. ({issue}`9214`)
```

* Prevent failure when querying `system.materialized_views` and materialized view is dropped concurrently. ({issue}`9050`) 
* Fix incorrect results for aggregations that contain `DISTINCT` and `FILTER` clauses. ({issue}`9265`)
* Fix `ALTER SCHEMA ... SET AUTHORIZATION` to resolve roles using the catalog of the schema instead of the session catalog. ({issue}`9066`)
* Fix query failure when query contains a cast from `varchar` to a shorter `char`. ({issue}`9036`)
* Fix planning failure of `INSERT` statement when source table has hidden columns. ({issue}`9150`)
* Fix planning of recursive queries when the recursion, the base plan, or the recursion step plan produce duplicate outputs.

> * Fix failure when querying `optimizer_rule_stats` system table. ({issue}`8700`) 

Reviewer 5: ```suggestion
* Fix failure when querying [](optimizer_rule_stats) system table. ({issue}`8700`) 
```

Reviewer 5: Or we link only when feature is added?

Author: No, that's better.

* Split metrics that are reported in operator stats into `connectorMetrics` and `metrics`. ({issue}`9156`)

## Security

* Add `IN catalog` clause to `CREATE ROLE`, `DROP ROLE`, `GRANT ROLE`, `REVOKE ROLE`, and `SET ROLE` to specify 
  the target catalog of the statement instead of using the current session catalog. This change is necessary to 
  support system roles in the future. The old behavior of these statements can be restored by setting the 
  `deprecated.legacy-catalog-roles` config property. ({issue}`9087`)

## BigQuery connector

* Support case insensitive name matching for BigQuery views. ({issue}`9164`)
* Change type mapping of BigQuery `datetime` from `timestamp(3)` to `timestamp(6)` in Trino. ({issue}`9052`)
* Change type mapping of BigQuery `time` from `time with time zone` to `time(6)` in Trino. ({issue}`9052`)
* Change type mapping of BigQuery `timestamp` from `timestamp(3) with time zone` to `timestamp(6) with time zone` in Trino. ({issue}`9052`)

> * Fix query failure for predicate pushdown to BigQuery `datetime` type. ({issue}`9005`)

Reviewer 5: ```suggestion
* Fix failure for queries where predicate on `datetime` column is pushed down to BigQuery. ({issue}`9005`)
```

* Fix incorrect results when using parameterized `numeric` type with non-default precision and scale. ({issue}`9060`)
* Fix query failure when accessing tables with unsupported data type columns. ({issue}`9086`)

> * Fix query failure when pushing down predicates to BigQuery on `float64` columns with `+infinity` or `-infinity` value. ({issue}`9122`)

Reviewer 5: ```suggestion
* Fix failure for queries where predicate on `float64` column, involving `+infinity` or `-infinity` values, is pushed down to BigQuery. ({issue}`9122`)
```

## Cassandra connector

* Change minimum number of speculative executions from 2 to 1. ({issue}`9096`)

## Hive connector

* Support reading Parquet timestamp stored with millisecond or microsecond precision as `INT64` with 
  logical type annotations when Hive timestamp precision is `NANOSECONDS`. ({issue}`9139`)
* Support reading Parquet timestamp stored as `INT64` with nanosecond precision. ({issue}`9188`)
* Support writing Parquet timestamp stored as `INT64` with nanosecond precision when experimental Parquet writer is enabled.
  To use this, the Hive timestamp precision should be set to `NANOSECONDS`. ({issue}`9188`)
* Support loading of s3 mappings via http/s url. The `hive.s3.security-mapping.config-file property` can now either point to a local file or a URL. ({issue}`6210`) 

> * Allow reads on table if they are bucketed on unsupported data type. ({issue}`8986`)

Reviewer 5: ```suggestion
* Allow reading from tables bucketed on a column that uses type for which bucketing is not natively supported by Trino. Writing to such tables is still not allowed. ({issue}`8986`)
```

* Extend set of statistics reported by JMX to include metrics for calls made to the Glue statistics API. ({issue}`9100`) 

> * Limit maximum file size generated by write operations to 1 GB by default. The limit is not exact and is applied on a best-effort basis. 
>   The limit can be set with the `target_max_file_size` session property, or the `hive.target-max-file-size` configuration property. ({issue}`7991`)

Reviewer 5: ```suggestion
* Limit maximum file size generated by write operations to 1 GB by default. The limit is not exact and is applied on a best-effort basis. 
  The limit can be set with the `hive.target-max-file-size` configuration property, or the `target_max_file_size` session property. ({issue}`7991`)
```

> * Optionally specify the schema which enforces that a filter on a partition key be present in the query for
>   its table. This can be specified by setting the ``hive.query-partition-filter-required-schemas`` config property or 
>   the ``query_partition_filter_required_schemas`` session property. ({issue}`9106`) 

Reviewer 5: ```suggestion
* Allow specifying the list of schemas for which Trino will enforce that queries use a filter on partition keys for source tables.
The list can be specified using the `hive.query-partition-filter-required-schemas`, or the `query_partition_filter_required_schemas` session property.
The list is taken into consideration only if the `hive.query-partition-filter-required` configuration property or the `query_partition_filter_required` session property is set to `true`. ({issue}`9106`) 
```

* Fix failure when writing Parquet files with Snappy compression on ARM64. ({issue}`9148`)
* Fix performance regression when reading Parquet files that contain dictionaries. ({issue}`9161`)
* Fix incorrect table locking in Hive metastore when querying transactional and not partitioned table. ({issue}`9070`) 
* Fix `ArrayIndexOutOfBoundsException` when using the experimental Parquet writer. ({issue}`5518`)

## Iceberg connector

* Fix the mapping of nested fields between table metadata and Parquet file metadata. This
  enables evolution of `row` typed columns for Iceberg tables stored in Parquet. ({issue}`9124`)
* Fix failure for queries filtering on columns with array, map, or row type. ({issue}`8822`)
* Fix `ArrayIndexOutOfBoundsException` when writing Parquet data. ({issue}`5518`)

## Pinot connector

* Support pushdown of filters on `VARBINARY` columns to Pinot. ({issue}`9180`)
* Fix incorrect results for queries that contain aggregations and `IN` and `NOT IN` filters over `varchar` columns. ({issue}`9133`)

> * Fix failure for queries on `real` or `double` columns having `+Infinity` or `-Infinity` values. ({issue}`9180`)

Reviewer 5: ```suggestion
* Fix failure for queries with filters on `real` or `double` columns having `+Infinity` or `-Infinity` values. ({issue}`9180`)
```

# Release 362 (17 Sep 2021)

## General

* Add support for {func}`listagg`. ({issue}`4835`)
* Improve join performance. ({issue}`8974`)
* Improve performance of queries that contain `DISTINCT` clause. ({issue}`8967`, {issue}`9194`)
* Improve query performance by reducing CPU overhead of repartitioning data across worker nodes. ({issue}`9102`)

> * Limit the number of backend SPI metadata calls when `java.sql.DatabaseMetaData` API is used to 

Reviewer 6: This is talking about implementation details. What is the user impact?

Reviewer 5: Let's make it:
```
Better performance of listing table columns via `java.sql.DatabaseMetaData` API when the filter is used on schema name.
```

* Fix failure for queries where predicate on `datetime` column is pushed down to BigQuery. ({issue}`9005`)
* Fix incorrect results when using parameterized `numeric` type with non-default precision and scale. ({issue}`9060`)
* Fix query failure when accessing tables with unsupported data type columns. ({issue}`9086`)
* Fix failure for queries where predicate on `float64` column, involving `+infinity` or 
  `-infinity` values, is pushed down to BigQuery. ({issue}`9122`)

## Cassandra connector

* Change minimum number of speculative executions from 2 to 1. ({issue}`9096`)

## Hive connector

* Support reading Parquet timestamp stored with millisecond or microsecond precision as `INT64` with 
  logical type annotations when Hive timestamp precision is `NANOSECONDS`. ({issue}`9139`)
* Support reading Parquet timestamp stored as `INT64` with nanosecond precision. ({issue}`9188`)
* Support writing Parquet timestamp stored as `INT64` with nanosecond precision when experimental Parquet writer is enabled.
  To use this, the Hive timestamp precision should be set to `NANOSECONDS`. ({issue}`9188`)

> * Support loading of s3 mappings via http/s url. The `hive.s3.security-mapping.config-file property` can now either point to a local file or a URL. ({issue}`6210`) 

Reviewer 6: Capitalize S3 and HTTP
